### PR TITLE
Add new partitioner script, which can do job on first boot

### LINKED ISSUE #1240: Support Docker inc & Redhat Docker distributions
Redhat docker tends to lag the official packages and has unresolved bugs. If we could create  common systemd overrides, we could have the same config for the different packages:

Docker Inc: 
```
[Unit]
Description=Docker Application Container Engine
Documentation=https://docs.docker.com
After=network.target docker.socket
Requires=docker.socket

[Service]
Type=notify
ExecStart=/usr/bin/docker daemon -H fd://
MountFlags=slave
LimitNOFILE=1048576
LimitNPROC=1048576
LimitCORE=infinity
TimeoutStartSec=0

[Install]
WantedBy=multi-user.target
```

Redhat:
```
[Unit]
Description=Docker Application Container Engine
Documentation=http://docs.docker.com
After=network.target
Wants=docker-storage-setup.service

[Service]
Type=notify
EnvironmentFile=-/etc/sysconfig/docker
EnvironmentFile=-/etc/sysconfig/docker-storage
EnvironmentFile=-/etc/sysconfig/docker-network
Environment=GOTRACEBACK=crash
ExecStart=/usr/bin/docker daemon $OPTIONS \
          $DOCKER_STORAGE_OPTIONS \
          $DOCKER_NETWORK_OPTIONS \
          $ADD_REGISTRY \
          $BLOCK_REGISTRY \
          $INSECURE_REGISTRY
LimitNOFILE=1048576
LimitNPROC=1048576
LimitCORE=infinity
MountFlags=slave
TimeoutStartSec=1min
Restart=on-failure

[Install]
WantedBy=multi-user.target

```

### PR DESCRIPTION
**Pull request by avnik**
- [x] Installs cleanly on a fresh build of most recent master branch
- [ ] Upgrades cleanly from the most recent release
- [x] Updates documentation relevant to the changes
- [x] Rebases cleanly onto the latest master

Fixes #1240

### CODE REVIEW

**Comment by avnik** [comment]
Add docker-io's docker support from @stevendborrelli

File: roles/docker/files/docker.repo [hunk 1]
```diff
@@ -0,0 +1,6 @@
+[dockerrepo]
```
Comment (stevendborrelli) [inline roles/docker/files/docker.repo]: There are 2 docker repos in the PR.

File: roles/lvm/files/mantl-storage-setup.py [hunk 2]
```diff
@@ -0,0 +1,296 @@
+#!/usr/bin/env python
```
Comment (stevendborrelli) [inline roles/lvm/files/mantl-storage-setup.py]: I thought this was moved to a package.

File: roles/docker/tasks/main.yml [hunk 3]
```diff
@@ -9,105 +9,38 @@
...
-    state: mounted
-    name: "{{ docker_volume_mount }}"
-    fstype: "{{ docker_volume_fs_type }}"
-    src: "{{ docker_volume_device }}"
-    dump: 1
-    passno: 2
-  when: docker_configure_lvm|bool
+  yum:
+    name: "{{ item }}"
+    state: absent
```
Comment (stevendborrelli) [inline roles/docker/tasks/main.yml]: Do we need to stop/start docker during the removal/installation process?

File: roles/docker/files/docker.repo (hunk 1 above)
Comment (avnik) [inline roles/docker/files/docker.repo]: Correct, I'll remove mine

File: roles/docker/tasks/main.yml (hunk 3 above)
Comment (avnik) [inline roles/docker/tasks/main.yml]: Manual need restart on upgrade.
Currently docker have started=yes in ansible, and it need to be switchable off to packer

File: roles/lvm/files/mantl-storage-setup.py (hunk 2 above)
Comment (avnik) [inline roles/lvm/files/mantl-storage-setup.py]: Is a local copy to debug and development purposes,   controlable with  debug_storage_setup=True in ansible

**Comment by stevendborrelli** [comment]
Documentation needs to be updated before this can go in.

**Comment by siddharthist** [comment]
We should update the Packer image after this is merged.

File: roles/lvm/files/mantl-storage-setup.rst [hunk 4]
```diff
@@ -0,0 +1,131 @@
+mantl-storage-setup script
+==========================
+
+Is a new partitioner solution for Mantl.
+
+Order of execution
+------------------
+
+All configuration rules reside in ``/etc/mantl/filesystems.d``, and processed
```
Comment (siddharthist) [inline roles/lvm/files/mantl-storage-setup.rst]: Maybe we could include an example configuration?

File: roles/lvm/tasks/debug.yml [hunk 5]
```diff
@@ -0,0 +1,33 @@
+# This rules executed only when debug_storage_setup == true
```
Comment (siddharthist) [inline roles/lvm/tasks/debug.yml]: Which shouldn't happen in production. Should we remove this from the PR as well (since this is already well tested)?

**Comment by siddharthist** [comment]
@avnik outlined these testing criteria: 
 - [ ] clean install
 - [ ] fresh install in both overlay+xfs and devicemapper
 - [ ] upgrade from 0.5.1 with devicemapper
 - [ ] downgrade docker-io's docker to RH one
 - [ ] upgrade from current master in both overlay+xfs and devicemapper

File: roles/docker/files/docker.repo (hunk 1 above)
Comment (siddharthist) [inline roles/docker/files/docker.repo]: This has been done.

File: roles/lvm/files/mantl-storage-setup.py (hunk 2 above)
Comment (siddharthist) [inline roles/lvm/files/mantl-storage-setup.py]: The debugging stuff will be removed when this PR is ready to merge.

**Comment by ryane** [comment]
on a clean install on AWS, I get the following error on the ansible run:

```
TASK: [docker | enable docker] ************************************************
failed: [resching-aws-worker-04] => {"failed": true}
msg: Job for docker.service failed because a configured resource limit was exceeded. See "systemctl status docker.service" and "journalctl -xe" for details.

failed: [resching-aws-worker-03] => {"failed": true}
msg: Job for docker.service failed because a configured resource limit was exceeded. See "systemctl status docker.service" and "journalctl -xe" for details.

failed: [resching-aws-worker-02] => {"failed": true}
msg: Job for docker.service failed because a configured resource limit was exceeded. See "systemctl status docker.service" and "journalctl -xe" for details.

failed: [resching-aws-worker-01] => {"failed": true}
msg: Job for docker.service failed because a configured resource limit was exceeded. See "systemctl status docker.service" and "journalctl -xe" for details.

failed: [resching-aws-control-01] => {"failed": true}
msg: Job for docker.service failed because a configured resource limit was exceeded. See "systemctl status docker.service" and "journalctl -xe" for details.

failed: [resching-aws-edge-01] => {"failed": true}
msg: Job for docker.service failed because a configured resource limit was exceeded. See "systemctl status docker.service" and "journalctl -xe" for details.

failed: [resching-aws-control-02] => {"failed": true}
msg: Job for docker.service failed because a configured resource limit was exceeded. See "systemctl status docker.service" and "journalctl -xe" for details.

failed: [resching-aws-edge-02] => {"failed": true}
msg: Job for docker.service failed because a configured resource limit was exceeded. See "systemctl status docker.service" and "journalctl -xe" for details.

failed: [resching-aws-control-03] => {"failed": true}
msg: Job for docker.service failed because a configured resource limit was exceeded. See "systemctl status docker.service" and "journalctl -xe" for details.


FATAL: all hosts have already failed -- aborting
```

`systemctl status docker.service`:

```
[centos@ip-10-1-1-9 ~]$ sudo systemctl status docker.service
● docker.service - Docker Application Container Engine
   Loaded: loaded (/usr/lib/systemd/system/docker.service; enabled; vendor preset: disabled)
  Drop-In: /etc/systemd/system/docker.service.d
           └─10-options.conf, 12-network-options.conf, 14-storage-options.conf, 20-ExecStart.conf
   Active: failed (Result: resources)
     Docs: https://docs.docker.com

Apr 12 12:05:27 ip-10-1-1-9.ec2.internal systemd[1]: Failed to load environment files: No such file or directory
Apr 12 12:05:27 ip-10-1-1-9.ec2.internal systemd[1]: docker.service failed to run 'start' task: No such file or directory
Apr 12 12:05:27 ip-10-1-1-9.ec2.internal systemd[1]: Failed to start Docker Application Container Engine.
Apr 12 12:05:27 ip-10-1-1-9.ec2.internal systemd[1]: Unit docker.service entered failed state.
Apr 12 12:05:27 ip-10-1-1-9.ec2.internal systemd[1]: docker.service failed.
Apr 12 12:05:27 ip-10-1-1-9.ec2.internal systemd[1]: Starting Docker Application Container Engine...
```

any idea? @avnik @stevendborrelli

**Comment by ryane** [comment]
same thing on an upgrade from 1.0.3

**Comment by avnik** [comment]
@ryane I guess it because we not update package with mantl-storage-setup, I'll do this atm (because I think code of partitioner enough stable).
Pass `debug_storage_setup=True` to ansible-playbook is another option

File: roles/docker/defaults/main.yml [hunk 6]
```diff
@@ -1,12 +1,41 @@
...
 
+# Docker log options: See https://docs.docker.com/engine/admin/logging/overview/
+docker_log_driver: journald
+
+# Docker options
+docker_options:
+  - "--dns {{ private_ipv4 }}"
+  - "--log-driver={{ docker_log_driver }}"
+  - "{% if docker_selinux_enabled %}--selinux-enabled {% endif %}"
+  - "{% if kube_build is defined %}--dns-search {{ cluster_name }}{% endif %}"
```
Comment (ryane) [inline roles/docker/defaults/main.yml]: is kube_build ever set anywhere?

**Comment by ryane** [comment]
successful clean build on GCE. However, I am seeing some unexpected docker errors. I restarted the nginx-consul service on a node and received the following:

```
Apr 27 22:54:11 resching-gce-control-03 systemd[1]: Stopping nginx-consul...
Apr 27 22:54:11 resching-gce-control-03 docker[2012]: nginx-consul
Apr 27 22:54:11 resching-gce-control-03 docker[7983]: Error response from daemon: Unable to remove filesystem for 566be3da7914736c38e32db59f2a04599c8fe76191fee610d7a67e71908f9847: remove /var/lib/docker/conta
iners/566be3da7914736c38e32db59f2a04599c8fe76191fee610d7a67e71908f9847/shm: device or resource busy
Apr 27 22:54:11 resching-gce-control-03 systemd[1]: nginx-consul.service: main process exited, code=exited, status=137/n/a
Apr 27 22:54:11 resching-gce-control-03 systemd[1]: Unit nginx-consul.service entered failed state.
Apr 27 22:54:11 resching-gce-control-03 systemd[1]: nginx-consul.service failed.
Apr 27 22:54:11 resching-gce-control-03 systemd[1]: Starting nginx-consul...
Apr 27 22:54:11 resching-gce-control-03 docker[2024]: Error response from daemon: Unable to remove filesystem for 566be3da7914736c38e32db59f2a04599c8fe76191fee610d7a67e71908f9847: remove /var/lib/docker/conta
iners/566be3da7914736c38e32db59f2a04599c8fe76191fee610d7a67e71908f9847/shm: device or resource busy
Apr 27 22:54:11 resching-gce-control-03 docker[2028]: 1.2: Pulling from ciscocloud/nginx-consul
...
Apr 27 22:55:33 resching-gce-control-03 docker[2875]: Status: Image is up to date for ciscocloud/nginx-consul:1.2
Apr 27 22:55:33 resching-gce-control-03 systemd[1]: Started nginx-consul.
Apr 27 22:55:33 resching-gce-control-03 docker[2880]: /usr/bin/docker: Error response from daemon: Conflict. The name "/nginx-consul" is already in use by container 566be3da7914736c38e32db59f2a04599c8fe76191fee610d7a67e71908f9847. You have to remove (or rename) that container to be able to reuse that name..
Apr 27 22:55:33 resching-gce-control-03 docker[2880]: See '/usr/bin/docker run --help'.
```

The container is marked as `Dead` and cannot be removed:

```
566be3da7914        ciscocloud/nginx-consul:1.2                           "/scripts/launch.sh"     31 minutes ago      Dead                                                                                                                          nginx-consul

$ sudo docker rm 566be3da7914
Error response from daemon: Unable to remove filesystem for 566be3da7914736c38e32db59f2a04599c8fe76191fee610d7a67e71908f9847: remove /var/lib/docker/containers/566be3da7914736c38e32db59f2a04599c8fe76191fee610d7a67e71908f9847/shm: device or resource busy
```

**Comment by ryane** [comment]
the issue doesn't seem to be directly related to the code in this pr but just may be a bug with the current version of docker. There are some similar issues open:

- https://github.com/docker/docker/issues/16982
- https://github.com/docker/docker/issues/21704
- https://github.com/docker/docker/issues/18509
- https://github.com/docker/docker/issues/20268

After discussing, we will merge this and workaround the immediate problem in #1390. We will open new issues for any other that may crop up.

**Pull request merged**